### PR TITLE
intel_adsp: hd audio: Add Power Management and Fix Initialization Issues for Intel HDA DAI and DMA Drivers

### PR DESCRIPTION
**Pull request by tmleman**
This pull request includes several improvements to the Intel HDA DAI and HDA DMA drivers. It adds power management support to the HDA DAI driver, ensuring balanced power management calls and correct power domain assignments. The HDA DMA driver is refactored to correct power management initialization and power domain assignments, ensuring proper power state management and reducing power consumption. Additionally, it addresses the invalid initialization sequence and register use in dma_config by ensuring channels are initialized during device initialization. These changes enhance system stability and ensure proper initialization and power management.

### CODE REVIEW

**Review by kv2019i — COMMENTED** [review]
Looks like a good improvement again. Once question inline.

**Comment by lyakh** [comment]
@tmleman do we have any tests that are currently failing and that this PR fixes? At least a description of such cases could be helpful

File: drivers/dai/intel/hda/hda.c [hunk 1]
```diff
@@ -122,11 +151,14 @@ static const struct dai_driver_api dai_intel_hda_api_funcs = {
...
+	PM_DEVICE_DT_INST_DEFINE(n, hda_pm_action);		\
+								\
 	DEVICE_DT_INST_DEFINE(n,				\
-			NULL, NULL,				\
+			hda_init, PM_DEVICE_DT_INST_GET(n),	\
 			&dai_intel_hda_data_##n,		\
 			&dai_intel_hda_config_##n,		\
-			POST_KERNEL, 32,			\
+			POST_KERNEL,				\
+			CONFIG_DAI_INIT_PRIORITY,		\
```
[the remark this answers is not in the document]
Comment (tmleman) [reply drivers/dai/intel/hda/hda.c:161]: The change in priority is forced by adding a dependency on the power domain:

```txt
Generating files from /home/tmleman/work/repos/converged/build-ptl/zephyr/zephyr.elf for board: intel_adsp
ERROR: Device initialization priority validation failed, the sequence of initialization calls does not match the devicetree dependencies.
ERROR: /hdas/hda@0 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+4 < POST_KERNEL+23)
ERROR: /hdas/hda@1 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+5 < POST_KERNEL+23)
ERROR: /hdas/hda@2 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+6 < POST_KERNEL+23)
ERROR: /hdas/hda@3 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+7 < POST_KERNEL+23)
ERROR: /hdas/hda@4 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+8 < POST_KERNEL+23)
ERROR: /hdas/hda@5 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+9 < POST_KERNEL+23)
ERROR: /hdas/hda@6 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+10 < POST_KERNEL+23)
ERROR: /hdas/hda@7 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+11 < POST_KERNEL+23)
ERROR: /hdas/hda@8 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+12 < POST_KERNEL+23)
ERROR: /hdas/hda@9 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+13 < POST_KERNEL+23)
ERROR: /hdas/hda@a <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+14 < POST_KERNEL+23)
ERROR: /hdas/hda@b <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+15 < POST_KERNEL+23)
ERROR: /hdas/hda@c <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+16 < POST_KERNEL+23)
ERROR: /hdas/hda@d <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+17 < POST_KERNEL+23)
ERROR: /hdas/hda@e <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+18 < POST_KERNEL+23)
ERROR: /hdas/hda@f <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+19 < POST_KERNEL+23)
ERROR: /hdas/hda@10 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+20 < POST_KERNEL+23)
ERROR: /hdas/hda@11 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+21 < POST_KERNEL+23)
ERROR: /hdas/hda@12 <hda_init> is initialized before its dependency /soc/dfpmccu@71b00/io0_domain <pd_intel_adsp_init> (POST_KERNEL+22 < POST_KERNEL+23)
ninja: build stopped: subcommand failed.
```

Initially, I changed it to the priority that is set in SSP (42), but later I found this config. Its name indicates that it is dedicated for this purpose, but I have not yet fully tested these changes.

**Comment by tmleman** [comment]
> do we have any tests that are currently failing and that this PR fixes? At least a description of such cases could be helpful

@lyakh Failures occur in the internal CI with a wider scope of tests. They can be observed in tests with DMIC and SSP. The reproduction rate is less than 10% (I have never managed to reproduce it locally).
The problem is quite obvious when you look at it closely. The function `intel_adsp_hda_channels_init` must be called when we raise our device. In the current version, if the device is under a power domain, it will only execute during `dma_start`, but `dma_config` is called earlier and unfortunately touches HW registers. To be more precise, the function `intel_adsp_hda_set_buffer` returns `-EBUSY` because some registers have incorrect values at the start.
Related commit: `009815e985db5e19b4a91413d2d9da9b265a0c2a`

**Comment by tmleman** [comment]
Marking as DNM until I confirm that the change does not introduce regression in SOF.

**Comment by lyakh** [comment]
> > do we have any tests that are currently failing and that this PR fixes? At least a description of such cases could be helpful
> 
> @lyakh Failures occur in the internal CI with a wider scope of tests. They can be observed in tests with DMIC and SSP. The reproduction rate is less than 10% (I have never managed to reproduce it locally). The problem is quite obvious when you look at it closely. The function `intel_adsp_hda_channels_init` must be called when we raise our device. In the current version, if the device is under a power domain, it will only execute during `dma_start`, but `dma_config` is called earlier and unfortunately touches HW registers. To be more precise, the function `intel_adsp_hda_set_buffer` returns `-EBUSY` because some registers have incorrect values at the start. Related commit: `009815e985db5e19b4a91413d2d9da9b265a0c2a`

@tmleman thanks for the explanation!

**Comment by tmleman** [comment]
Rebase to enable testing in SOF CI: https://github.com/thesofproject/sof/pull/9714

**Comment by tmleman** [comment]
SOF CI did not show any regression. Failures at https://sof-ci.01.org/sofpr/PR9714/build9718/devicetest/index.html look like a platform issue.

**Review by serhiy-katsyuba-intel — APPROVED** [review]

**Review by LaurentiuM1234 — APPROVED** [review]

**Pull request merged**